### PR TITLE
Fixed pickling on SP

### PR DESCRIPTION
**Pull request by dkeeney**
This PR fixes the pickling on SP and adds saveToFile( filename ) and loadFromFile(filename).
Response to Issue #641

### CODE REVIEW

**Comment by dkeeney** [comment]
Now that I have something that works for pickling, I need to apply this to most of the other objects in the bindings.

**Comment by breznak** [comment]
https://travis-ci.org/htm-community/htm.core/builds/574537849#L2368 

Travis is still not happy

**Comment by dkeeney** [comment]
> Travis is still not happy

I typed in the name of the function incorrectly :-)

File: bindings/py/tests/algorithms/temporal_memory_test.py [hunk 1]
```diff
@@ -38,15 +38,51 @@ def testCompute(self):
     self.assertTrue( active.getSum() > 0 )
 
 
-  @pytest.mark.skipif(sys.version_info < (3, 6), reason="Fails for python2 with segmentation fault")
```
Comment (breznak) [inline bindings/py/tests/algorithms/temporal_memory_test.py]: isn't this still valid? (we don't test on py2, but someone still may run the repo on python2)

**Comment by dkeeney** [comment]
@breznak when you have some time can you give this a review again.

I want to point out that in bindings\py\tests\algorithms\temporal_memory_test.py, line 58
I marked this py unit test as skipped.  It is giving me an error that comes from Cereal. It is an error I have see before with JSON output and it may indicate some bad Cereal serialization code someplace.  I think this should be looked at with a different PR.  I will added an Issue for it. Issue #645 

Funny how when we add more tests we seem to find more bugs.

**Comment by dkeeney** [comment]
> Can you file an issue with pybind11? 

Let's be sure our code is good first.  I know that JSON is more sensitive to errors than the BINARY mode is.  This will take some time to investigate.  

>Is there a newer release we could try?

I checked.  They are still on 1.2.2 (Feb 2017)   Nothing has changed in the master branch in 3 years.  The error in JSON comes from rapidjson which is the third party parser. http://rapidjson.org/
Its latest release is Aug 2016 so nothing new.  But there are some more recent changes in their master.

**Comment by breznak** [comment]
> Nothing has changed in the master branch in 3 years.  

hope we didn't switch to a dead project/ I'm happy with a mature, stable codebases, but nothing in 3yrs sounds suspicious

**Comment by dkeeney** [comment]
I agree.  The buggy part is the JSON encoder and until rapidjson does a new release Cereal cannot either. (Note that this is Cereal and not pybind11 we are talking about)

**Comment by dkeeney** [comment]
I am still trying to understand how the build process works...
Can you tell me what GitHub nethook triggers we are using and perhaps what plugins we have?   I assume we are using nethooks to kick off the three CI's.  I don't want to change anything, just need to know what the parts are.

**Comment by breznak** [comment]
> Can you tell me what GitHub nethook triggers

I'm not sure what you mean, the CI are installed as "GH apps" and we have webhooks (all events allowed) which the CI react to: commit push, PR opened, etc.

**Pull request merged**